### PR TITLE
NaviPlanner:Support 108km/h lane maintenance, lane change and followi…

### PR DESCRIPTION
**Pull request by mfyanga**
…ng, u-turn
@yifeijiang  @vlin17 
1) Add priority fields to the reference line;
2) Navi Planner uses the reference line priority as the judgment condition for getting the target lane;
3) Add UpdateReferenceLinePriority to update reference line priority based on actual road conditions;
4) Add a change lane ros msg to let adc change lane;
5) Modify /apollo/modules/dreamview/conf/hmi.conf for no Navi Planning, no third_party_perception showing on dreamview under Navigation Mode;
6) Modify /apollo/scripts/navigation_planning.sh for loading the correct planning configuration file;

### CODE REVIEW

File: modules/planning/navi/decider/navi_obstacle_decider_test.cc [hunk 1]
```diff
@@ -65,11 +66,16 @@ TEST(NaviObstacleDeciderTest, ComputeNudgeDist1) {
   path_obstacles.SetPerceptionSlBoundary(obstacle_boundary);
   path_decision.AddPathObstacle(path_obstacles);
```
Comment (Capri2014) [inline modules/planning/navi/decider/navi_obstacle_decider_test.cc]: why set the linear velocity in vehicle_state a constant?

File: modules/common/adapters/adapter_gflags.cc [hunk 2]
```diff
@@ -29,6 +29,8 @@ DEFINE_string(localization_topic, "/apollo/localization/pose",
               "localization topic name");
 DEFINE_string(planning_trajectory_topic, "/apollo/planning",
               "planning trajectory topic name");
+DEFINE_string(change_lane_topic, "/apollo/change_lane",
```
Comment (lianglia-apollo) [inline modules/common/adapters/adapter_gflags.cc]: Why do we need a separate topic for change lane?

File: modules/planning/navi/decider/navi_obstacle_decider.cc [hunk 3]
```diff
@@ -152,50 +178,121 @@ int NaviObstacleDecider::ProcessPathObstacle(
...
+      }
+    }
+  }
+  ADEBUG << "last_nudge_dist : " << last_nudge_dist_
+         << "lat_dist : " << lat_dist << "actual_dist : " << actual_dist;
+  return actual_dist;
+}
+
+void NaviObstacleDecider::RecordLastNudgeDistance(const double nudge_dist) {
+  static unsigned int no_nudge_num = 0;
```
Comment (lianglia-apollo) [inline modules/planning/navi/decider/navi_obstacle_decider.cc]: need to reset static variables when starting (or stop) the planning module.

File: modules/planning/navi/decider/navi_obstacle_decider.cc [hunk 4]
```diff
@@ -152,50 +178,121 @@ int NaviObstacleDecider::ProcessPathObstacle(
...
+  double tolerance = config_.nudge_allow_tolerance();
+
+  if (nudge_dist > tolerance) {
+    if (last_nudge_dist_ < nudge_dist) {
+      last_nudge_dist_ = nudge_dist;
+    }
+    no_nudge_num = 0;
+  } else if (nudge_dist < -1.0 * tolerance) {
```
Comment (lianglia-apollo) [inline modules/planning/navi/decider/navi_obstacle_decider.cc]: why -1.0 * tolerance?

File: modules/planning/navi/decider/navi_obstacle_decider.cc [hunk 5]
```diff
@@ -152,50 +178,121 @@ int NaviObstacleDecider::ProcessPathObstacle(
...
+  }
+  if ((no_nudge_num >= config_.cycles_number()) &&
+      (fabs(last_nudge_dist_) > tolerance)) {
+    last_nudge_dist_ = 0.0;
+  }
+}
+
+void NaviObstacleDecider::NudgeSmoothingDeal(
+    const common::VehicleState& vehicle_state, double* nudge_dist) {
+  static unsigned int limit_speed_num = 0;
```
Comment (lianglia-apollo) [inline modules/planning/navi/decider/navi_obstacle_decider.cc]: same as the previous comment. We need to reset static variables to clean up all the status.

File: modules/planning/navi/decider/navi_obstacle_decider.cc [hunk 6]
```diff
@@ -152,50 +178,121 @@ int NaviObstacleDecider::ProcessPathObstacle(
...
+
+  if (vehicle_state.linear_velocity() < config_.max_allow_nudge_speed()) {
+    limit_speed_num = limit_speed_num + 1;
+  } else {
+    limit_speed_num = 0;
+  }
+  if (limit_speed_num < config_.cycles_number()) {
+    *nudge_dist = 0;
+  }
+  if (fabs(*nudge_dist) > config_.nudge_allow_tolerance()) {
```
Comment (lianglia-apollo) [inline modules/planning/navi/decider/navi_obstacle_decider.cc]: Please use std::fabs

File: modules/planning/navi/decider/navi_obstacle_decider.cc [hunk 7]
```diff
@@ -152,50 +178,121 @@ int NaviObstacleDecider::ProcessPathObstacle(
...
+  static unsigned int limit_speed_num = 0;
+  static unsigned int eliminate_clutter_num = 0;
+
+  if (vehicle_state.linear_velocity() < config_.max_allow_nudge_speed()) {
+    limit_speed_num = limit_speed_num + 1;
+  } else {
+    limit_speed_num = 0;
+  }
+  if (limit_speed_num < config_.cycles_number()) {
+    *nudge_dist = 0;
```
Comment (lianglia-apollo) [inline modules/planning/navi/decider/navi_obstacle_decider.cc]: nudge_dist is not checked for nullptr.

File: modules/planning/navi/decider/navi_obstacle_decider.cc [hunk 8]
```diff
@@ -152,50 +178,121 @@ int NaviObstacleDecider::ProcessPathObstacle(
...
-  // of the obstacle.
+  // Calculating the left and right nudgable distance according to the
+  // position of the obstacle.
   double left_nudge_obstacle = 0.0;
   double right_nudge_obstacle = 0.0;
   // Calculation of the number of current Lane obstacles
   obstacle_lat_dist_.clear();
-  *lane_obstacles_num = ProcessPathObstacle(obstacles, path_data_points,
-                                            path_decision, min_lane_width);
+  *lane_obstacles_num = 0;
```
Comment (lianglia-apollo) [inline modules/planning/navi/decider/navi_obstacle_decider.cc]: this seems to be unnecessary.

File: modules/planning/navi/decider/navi_obstacle_decider.cc [hunk 9]
```diff
@@ -224,21 +321,54 @@ double NaviObstacleDecider::GetNudgeDistance(
...
+
+  ADEBUG << "nudge_dist : " << nudge_dist
+         << " last_nudge_dist : " << last_nudge_dist;
+  if ((fabs(nudge_dist) > config_.nudge_allow_tolerance()) &&
+      (fabs(last_nudge_dist) < config_.nudge_allow_tolerance()) &&
+      (false == keep_nudge_flag)) {
+    count = 0;
+    keep_nudge_flag = true;
+  }
+  if (true == keep_nudge_flag) {
```
Comment (lianglia-apollo) [inline modules/planning/navi/decider/navi_obstacle_decider.cc]: if (keep_nudge_flag)

File: modules/planning/navi/decider/navi_obstacle_decider.cc [hunk 10]
```diff
@@ -224,21 +321,54 @@ double NaviObstacleDecider::GetNudgeDistance(
...
 
+void NaviObstacleDecider::KeepNudgePosition(const double nudge_dist,
+                                            int* lane_obstacles_num) {
+  static double last_nudge_dist = 0.0;
+  static size_t count = 0;
+  static bool keep_nudge_flag = false;
+
+  ADEBUG << "nudge_dist : " << nudge_dist
+         << " last_nudge_dist : " << last_nudge_dist;
+  if ((fabs(nudge_dist) > config_.nudge_allow_tolerance()) &&
```
Comment (lianglia-apollo) [inline modules/planning/navi/decider/navi_obstacle_decider.cc]: use std::fabs

File: modules/planning/navi/decider/navi_obstacle_decider.h [hunk 11]
```diff
@@ -76,45 +69,108 @@ class NaviObstacleDecider : public Task {
...
+   */
+  void NudgeSmoothingDeal(const common::VehicleState &vehicle_state,
+                          double *nudge_dist);
+  void KeepNudgePosition(const double nudge_dist, int *lane_obstacles_num);
+
  private:
+  NaviObstacleDeciderConfig config_;
   std::map<double, double> obstacle_lat_dist_;
   std::vector<std::tuple<std::string, double, double>> unsafe_obstacle_info_;
+  double last_nudge_dist_;
```
Comment (lianglia-apollo) [inline modules/planning/navi/decider/navi_obstacle_decider.h]: Please initialize private members

File: modules/planning/planning_base.cc [hunk 12]
```diff
@@ -260,8 +260,13 @@ Status PlanningBase::Plan(
 
   ADEBUG << "current_time_stamp: " << std::to_string(current_time_stamp);
 
-  last_publishable_trajectory_->PrependTrajectoryPoints(
-      stitching_trajectory.begin(), stitching_trajectory.end() - 1);
+  // Navi Panner doesn't need to stitch the last path planning
```
Comment (lianglia-apollo) [inline modules/planning/planning_base.cc]: Please use gflag when staring navi planner to control whether to stitch trajectory or not.

File: modules/planning/reference_line/reference_line_provider.cc [hunk 13]
```diff
@@ -432,7 +432,9 @@ bool ReferenceLineProvider::GetReferenceLinesFromRelativeMap(
                        LaneWaypoint(lane_ptr, path_point.s())},
           path_point.kappa(), path_point.dkappa());
     }
-    reference_line->emplace_back(ref_points.begin(), ref_points.end());
+    ReferenceLine ref_line(ref_points);
```
Comment (lianglia-apollo) [inline modules/planning/reference_line/reference_line_provider.cc]: We can set priority after emplaced_back the reference line.

**Comment by davidhopper2003** [comment]
@lianglia-apollo 
It is a temporary solution to change a lane with the help of a ROS topic. In the future, we will replace it with a decision algorithm. From a security perspective, this feature can be rejected.

File: modules/planning/navi/decider/navi_obstacle_decider.cc (hunk 3 above)
Comment (davidhopper2003) [reply modules/planning/navi/decider/navi_obstacle_decider.cc]: Yes, it should be.

File: modules/planning/reference_line/reference_line_provider.cc (hunk 13 above)
Comment (mfyanga) [reply modules/planning/reference_line/reference_line_provider.cc]: ok

File: modules/planning/planning_base.cc (hunk 12 above)
Comment (mfyanga) [reply modules/planning/planning_base.cc]: ok

File: modules/planning/navi/decider/navi_obstacle_decider.cc (hunk 8 above)
Comment (mfyanga) [reply modules/planning/navi/decider/navi_obstacle_decider.cc]: yes,it is unnecessary

File: modules/planning/navi/decider/navi_obstacle_decider.cc (hunk 9 above)
Comment (mfyanga) [reply modules/planning/navi/decider/navi_obstacle_decider.cc]: ok

File: modules/planning/navi/decider/navi_obstacle_decider.cc (hunk 7 above)
Comment (mfyanga) [reply modules/planning/navi/decider/navi_obstacle_decider.cc]: ok

File: modules/planning/navi/decider/navi_obstacle_decider.cc (hunk 6 above)
Comment (mfyanga) [reply modules/planning/navi/decider/navi_obstacle_decider.cc]: ok

File: modules/planning/navi/decider/navi_obstacle_decider.cc (hunk 5 above)
Comment (mfyanga) [reply modules/planning/navi/decider/navi_obstacle_decider.cc]: yes,it should be clean up

File: modules/common/adapters/adapter_gflags.cc (hunk 2 above)
Comment (lianglia-apollo) [reply modules/common/adapters/adapter_gflags.cc]: @davidhopper2003 We discussed internally and suggest making a topic for the remote control of ego vehicle. The change lane signal can be one of the signals that belongs to this topic.

Comment (yifeijiang) [reply modules/common/adapters/adapter_gflags.cc]: @davidhopper2003 @mfyanga Could you change the topic name from "/apollo/change_lane" to "/apollo/planning/pad" as we would like to use this topic for other manual inputs besides change lane for planning module?

**Comment by mfyanga** [comment]
I will create a new pull a request.

**Pull request closed without merging**